src/lang/std/artifactGraph.ts

```diff
@@ -0,0 +1,645 @@
+  codeRef: CommonCommandProperties
+}
+
+interface _WallArtifact {
+  type: 'wall'
+  segId: string
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+interface _CapArtifact {
+  type: 'cap'
+  subType: 'start' | 'end'
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+
+interface _ExtrudeEdge {
+  type: 'extrudeEdge'
+  segId: string
+  extrusionId: string
+  edgeId: string
+}
+
+/** A edgeCut is a more generic term for both fillet or chamfer */
+interface _EdgeCut {
+  type: 'edgeCut'
+  subType: 'fillet' | 'chamfer'
+  consumedEdgeId: string
+  edgeIds: Array<string>
+  surfId: string
+  codeRef: CommonCommandProperties
+}
+
+interface EdgeCutEdge {
+  type: 'edgeCutEdge'
+  edgeCutId: string
+  surfId: string
+}
+
+export type Artifact =
+  | _PlaneArtifact
+  | _PathArtifact
+  | _SegmentArtifact
+  | _ExtrusionArtifact
+  | _WallArtifact
+  | _CapArtifact
+  | _ExtrudeEdge
+  | _EdgeCut
+  | EdgeCutEdge
+  | _solid2D
+
+export type ArtifactGraph = Map<string, Artifact>
+
+export type EngineCommand = Models['WebSocketRequest_type']
+
+type OkWebSocketResponseData = Models['OkWebSocketResponseData_type']
+
+export interface ResponseMap {
+  [commandId: string]: OkWebSocketResponseData
+}
+export interface OrderedCommand {
+  command: EngineCommand
+  range: SourceRange
+}
+
+export function createArtifactGraph({
+  orderedCommands,
+  responseMap,
+  ast,
+}: {
+  orderedCommands: Array<OrderedCommand>
+  responseMap: ResponseMap
+  ast: Program
+}) {
+  const myMap = new Map<string, Artifact>()
+  let currentPlaneId = ''
+
+  const getArtifact = (id: string) => myMap.get(id)
+  orderedCommands.forEach((orderedCommand) => {
+    if (orderedCommand.command?.type === 'modeling_cmd_req') {
+      if (orderedCommand.command.cmd.type === 'enable_sketch_mode') {
+        currentPlaneId = orderedCommand.command.cmd.entity_id
+      }
+      if (orderedCommand.command.cmd.type === 'sketch_mode_disable') {
+        currentPlaneId = ''
+      }
+    }
+    const modArr = getArtifactsToUpdate({
+      orderedCommand,
+      responseMap,
+      getArtifact,
+      currentPlaneId,
+      ast,
+    })
+    modArr.forEach(({ id, artifact }) => {
+      const oldArtifact = myMap.get(id)
+      if (oldArtifact?.type !== artifact.type) {
+        myMap.set(id, artifact)
+        return
+      }
+      const mergedArtifact = { ...oldArtifact, ...artifact }
+      Object.entries(artifact).forEach(([propName, value]) => {
+        const otherValue = (oldArtifact as any)[propName]
+        if (Array.isArray(value) && Array.isArray(otherValue)) {
+          ;(mergedArtifact as any)[propName] = [
+            ...new Set([...otherValue, ...value]),
+          ]
+        }
+      })
+      myMap.set(id, mergedArtifact)
+    })
+  })
+  return myMap
+}
+
+export function getArtifactsToUpdate({
+  orderedCommand: { command, range },
+  getArtifact,
+  responseMap,
+  currentPlaneId,
```

Comment:
`currentPlaneId` seems odd to me, it seems quite a specific thing to have to pass in for something as general as the ArtifactGraph no? Or is it because every CAD scene in some way start from a root that is a plane?

Reply:
So different paths/sketchs can start from different planes, but we can't link to the specific plane when only looking at one command at a time (which is what I was hoping for this function)

The reason why is because it depends on a sequence of commands, i.e.

- enable sketch mode (plane id)
- start_path (use the plane Id from the above)
- disable sktech mode


So this multi-command state I keep track of in the loop in `createArtifactGraph` to pass to this function.
It's only needed for the `start_path` command, but I kinda also need to pass it in.

We we can get the engine to update `start_path` to take a plane Id instead of it relying on implicit state, we can remove this.